hazelcast/src/test/java/com/hazelcast/client/alto/ClientAltoTest.java

```diff
@@ -0,0 +1,378 @@
+
+        assertTrueEventually(() -> assertEquals(1, connections.size()));
+        assertClientConnectsAllAltoPortsEventually(connections, config.getAltoConfig().getEventloopCount());
+    }
+
+    @Test
+    public void testClientRoutesPartitionBoundRequestsToAltoConnections() {
+        Config config = getMemberConfig();
+        HazelcastInstance instance1 = Hazelcast.newHazelcastInstance(config);
+        HazelcastInstance instance2 = Hazelcast.newHazelcastInstance(config);
+        warmUpPartitions(instance1, instance2);
+
+        HazelcastInstance client = HazelcastClient.newHazelcastClient(getClientConfig());
+        IMap<String, Integer> map = client.getMap(randomMapName());
+
+        ClientConnectionManager connectionManager = getConnectionManager(client);
+        Collection<Connection> connections = connectionManager.getActiveConnections();
+
+        assertClientConnectsAllAltoPortsEventually(connections, config.getAltoConfig().getEventloopCount());
+
+        // Wait until the partition table is received on the client-side
+        PartitionService partitionService = client.getPartitionService();
+        assertTrueEventually(() -> assertFalse(partitionService.getPartitions().isEmpty()));
+
+        int partitionCount = partitionService.getPartitions().size();
+        for (int i = 0; i < partitionCount; i++) {
+            String key = generateKeyForPartition(instance1, i);
+            long currentTimeMillis = System.currentTimeMillis();
+            map.put(key, i);
+
+            UUID ownerUuid = partitionService.getPartition(key).getOwner().getUuid();
+            TcpClientConnection partitionOwner = (TcpClientConnection) connectionManager.getConnection(ownerUuid);
+            assertNotNull(partitionOwner);
+
+            Channel[] altoChannels = partitionOwner.getAltoChannels();
+            assertNotNull(altoChannels);
+
+            Channel altoChannel = altoChannels[i % altoChannels.length];
+            assertFalse(altoChannel.isClosed());
+            assertTrue(altoChannel.lastWriteTimeMillis() >= currentTimeMillis);
+        }
+    }
+
+    @Test
+    public void testClientRoutesNonPartitionBoundRequestsToClassicConnections() {
+        Config config = getMemberConfig();
+        Hazelcast.newHazelcastInstance(config);
+
+        HazelcastInstance client = HazelcastClient.newHazelcastClient(getClientConfig());
+        IMap<String, Integer> map = client.getMap(randomMapName());
+
+        ClientConnectionManager connectionManager = getConnectionManager(client);
+        Collection<Connection> connections = connectionManager.getActiveConnections();
+
+        assertClientConnectsAllAltoPortsEventually(connections, config.getAltoConfig().getEventloopCount());
+
+        long currentTimeMillis = System.currentTimeMillis();
+        map.size();
+        ClientConnection connection = connectionManager.getRandomConnection();
+        assertTrue(connection.lastWriteTimeMillis() >= currentTimeMillis);
```

Comment:
maybe check it is not routed to alto channel?

Reply:
There is no reliable way of doing it. We are going to implement heartbeats for Alto channels, for example, then the last write time for the channel might be updated and the test might fail. I can do some tricks when we convert this test to mock network in the future